File: tests/samples/vulnerable.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"unsafe"
)

func vulnerable_function_1(userInput string) {
	// Rule 1: exec.Command with user input (Command Injection)
	cmd := exec.Command("bash", "-c", userInput)
	cmd.Run()
}

func vulnerable_function_2() {
	// Rule 2: unsafe.Pointer Usage
	var x int = 10
	ptr := unsafe.Pointer(&x)
	_ = ptr
}

func vulnerable_function_3() {
	// Rule 3: Ignored Error Returns
	f, _ := os.Open("secret.txt")
	defer f.Close()
	
	_, err := os.Create("test.txt")
}

func vulnerable_function_4(userFormat string) {
	// Rule 4: fmt.Sprintf Format String
	res := fmt.Sprintf(userFormat)
	fmt.Printf(userFormat)
}

func vulnerable_function_5() {
	// Rule 5: os.Args Without Validation
	arg := os.Args[1]
	fmt.Println(arg)
}

func vulnerable_function_6() {
	// Rule 6: Hardcoded Secrets
	apiKey := "AIzaSyDxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
	password := "superSecret123!"
	secretKey = "1234567890abcdef1234567890abcdef"
}

func main() {
	vulnerable_function_5()
	vulnerable_function_1(os.Args[1])
	vulnerable_function_2()
	vulnerable_function_3()
	vulnerable_function_4(os.Args[1])
	vulnerable_function_6()
}

File: tests/samples/vulnerable.py
import os
import subprocess
import pickle
import sys

def vulnerable_function_1(user_input):
    # Rule 1: eval() / exec()
    eval(user_input)
    exec(user_input)
    eval("print(" + user_input + ")")
    exec("import os; os.system('" + user_input + "')")

def vulnerable_function_2(user_data):
    # Rule 2: Unsafe Deserialization
    pickle.loads(user_data)
    pickle.load(open(user_data, 'rb'))
    import pickle as pkl
    pkl.loads(user_data)

def vulnerable_function_3(cmd):
    # Rule 3: subprocess shell=True
    subprocess.call(cmd, shell=True)
    subprocess.Popen(cmd, shell=True)
    subprocess.run(cmd, shell = True)
    subprocess.getoutput(cmd)

def vulnerable_function_4(cmd):
    # Rule 4: os.system()
    os.system(cmd)
    os.popen(cmd)
    os.system("ping " + cmd)
    os.popen("ls " + cmd)

def vulnerable_function_5(filename):
    # Rule 5: Path Traversal via open()
    f = open(filename, 'r')
    data = f.read()
    f.close()
    
    with open(sys.argv[1] + filename) as file:
        pass

def vulnerable_function_6():
    # Rule 6: Hardcoded Secrets
    password = "SuperSecretPassword123!"
    api_key = "AIzaSyD..._something_secret"
    secret = "my_secret_token"
    aws_secret_key = "AKIAIOSFODNN7EXAMPLE"
    auth_token = "abcd1234efgh5678"

if __name__ == "__main__":
    vulnerable_function_1(sys.argv[1])
    vulnerable_function_2(sys.argv[1].encode())
    vulnerable_function_3(sys.argv[1])
    vulnerable_function_4(sys.argv[1])
    vulnerable_function_5(sys.argv[1])
    vulnerable_function_6()